# Data preparation

This code is to prepare dataset for modelling a classifier to discriminate emotions using MSP dataset: [Link](https://ecs.utdallas.edu/research/researchlabs/msp-lab/MSP-Podcast.html).

The prerequisite of this notebook is to segment the audio files by following the guidance provided in the MSP dataset. `audio_crop.py` file segmented big chunk audio files into 2-4 audio segments and also 8 seconds in average length audio files for train, test, and validation dataset.

This file is just for practice to extract audio features and put all emotion data into csv files (train, test, validation)

In [2]:
import os
import json
import pandas as pd
from util.audio_crop import file_labels
import soundfile as sf
import opensmile


In [3]:
segment_path = os.path.relpath('./MSP Data/Time Labels/segments.json')
f = open(segment_path, 'r')
timing_data = json.load(f)

max_time = 0
for key in timing_data:
    start = timing_data[key]['Start_Time']
    end = timing_data[key]['End_Time']
    max_time = max(max_time, end-start)
max_time

193.61999999999992

In [2]:
# save paths
arouse_path = os.path.relpath('./MSP Data/Annotations/Arousal')
dominance_path = os.path.relpath('./MSP Data/Annotations/Dominance')
valence_path = os.path.relpath('./MSP Data/Annotations/Valence')
train_path = os.path.relpath('./Data/train')
validation_path = os.path.relpath('./Data/validation')
test_path = os.path.relpath('./Data/test')
data_path = os.path.relpath('./Data')

In [3]:
train = pd.DataFrame()
test = pd.DataFrame()
validation = pd.DataFrame()

In [26]:
train_files, validation_files, test_files = file_labels()

Number of training labels: 40
Number of validation labels: 14
Number of testing labels: 20


In [5]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.eGeMAPSv02,
    feature_level=opensmile.FeatureLevel.Functionals,
    num_workers=4
)


In [58]:
file_paths = []
for file in os.listdir(validation_path):
    file_paths.append(os.path.join(validation_path, file))
data = smile.process_files(file_paths)

/home/changhyun/workspace/ABI_research/env/lib/python3.8/site-packages/opensmile/core/smile.py:308: UserWarning: Segment too short, filling with NaN.
  warnings.warn(


In [59]:
# Errors in segment.json file that there are the same start and end time for MSP-PODCAST_0153 audio
data[data.isna().any(axis=1)]
# train

F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                      start  end                                                         
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                      start  end                                                              
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                     NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file                                      start  end                                                                
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                       NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file                                      start  end                                                                   
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                          NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file                                      start  end                                                                     
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                            NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file                                      start  end                                                                    
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                           NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file                                      start  end                                                                      
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                             NaN   

                                                                         ...  \
file                                      start  end                     ...   
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000  ...   

                                                                         slopeUV0-500_sma3nz_amean  \
file                                      start  end                                                 
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:0

In [60]:
data = data.reset_index(level=[1,2])
data

start                       end  \
file                                                                         
Data/validation/MSP-PODCAST_0744_0066.wav 0 days    0 days 00:00:16.630000   
Data/validation/MSP-PODCAST_0726_0024.wav 0 days    0 days 00:00:07.590000   
Data/validation/MSP-PODCAST_0726_0094.wav 0 days    0 days 00:00:05.980000   
Data/validation/MSP-PODCAST_0653_0154.wav 0 days 0 days 00:00:07.560062500   
Data/validation/MSP-PODCAST_1190_0058.wav 0 days    0 days 00:00:05.210000   
...                                          ...                       ...   
Data/validation/MSP-PODCAST_0726_0010.wav 0 days    0 days 00:00:15.910000   
Data/validation/MSP-PODCAST_1190_0008.wav 0 days           0 days 00:00:04   
Data/validation/MSP-PODCAST_1185_0019.wav 0 days 0 days 00:00:45.359954649   
Data/validation/MSP-PODCAST_1190_0043.wav 0 days    0 days 00:00:01.740000   
Data/validation/MSP-PODCAST_0396_2225.wav 0 days 0 days 00:00:01.029937500   

                                           F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                                                           
Data/validation/MSP-PODCAST_0744_0066.wav                          32.460304   
Data/validation/MSP-PODCAST_0726_0024.wav                          34.533546   
Data/validation/MSP-PODCAST_0726_0094.wav                          34.142075   
Data/validation/MSP-PODCAST_0653_0154.wav                          39.902229   
Data/validation/MSP-PODCAST_1190_0058.wav                          32.839245   
...                                                                      ...   
Data/validation/MSP-PODCAST_0726_0010.wav                          25.751215   
Data/validation/MSP-PODCAST_1190_0008.wav                          31.448648   
Data/validation/MSP-PODCAST_1185_0019.wav                          31.897440   
Data/validation/MSP-PODCAST_1190_0043.wav                          33.760132   
Data/validation/MSP-PODCAST_0396_2225.wav                          28.339310   

                                           F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                                                                
Data/validation/MSP-PODCAST_0744_0066.wav                                0.122227   
Data/validation/MSP-PODCAST_0726_0024.wav                                0.103023   
Data/validation/MSP-PODCAST_0726_0094.wav                                0.064815   
Data/validation/MSP-PODCAST_0653_0154.wav                                0.098032   
Data/validation/MSP-PODCAST_1190_0058.wav                                0.371941   
...                                                                           ...   
Data/validation/MSP-PODCAST_0726_0010.wav                                0.247039   
Data/validation/MSP-PODCAST_1190_0008.wav                                0.257237   
Data/validation/MSP-PODCAST_1185_0019.wav                                0.175576   
Data/validation/MSP-PODCAST_1190_0043.wav                                0.274083   
Data/validation/MSP-PODCAST_0396_2225.wav                                0.134931   

                                           F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                                                                    
Data/validation/MSP-PODCAST_0744_0066.wav                                   29.499130   
Data/validation/MSP-PODCAST_0726_0024.wav                                   32.856342   
Data/validation/MSP-PODCAST_0726_0094.wav                                   33.418037   
Data/validation/MSP-PODCAST_0653_0154.wav                                   37.958176   
Data/validation/MSP-PODCAST_1190_0058.wav                                   15.883025   
...                                                                               ...   
Data/validation/MSP-PODCAST_0726_0010.wav                                   22.775183   
Data/validation/MSP-PODCAST_1190_0008.wav                                   26.7

In [62]:
# train.dropna()
data = data.dropna()
data

start                       end  \
file                                                                         
Data/validation/MSP-PODCAST_0744_0066.wav 0 days    0 days 00:00:16.630000   
Data/validation/MSP-PODCAST_0726_0024.wav 0 days    0 days 00:00:07.590000   
Data/validation/MSP-PODCAST_0726_0094.wav 0 days    0 days 00:00:05.980000   
Data/validation/MSP-PODCAST_0653_0154.wav 0 days 0 days 00:00:07.560062500   
Data/validation/MSP-PODCAST_1190_0058.wav 0 days    0 days 00:00:05.210000   
...                                          ...                       ...   
Data/validation/MSP-PODCAST_0726_0010.wav 0 days    0 days 00:00:15.910000   
Data/validation/MSP-PODCAST_1190_0008.wav 0 days           0 days 00:00:04   
Data/validation/MSP-PODCAST_1185_0019.wav 0 days 0 days 00:00:45.359954649   
Data/validation/MSP-PODCAST_1190_0043.wav 0 days    0 days 00:00:01.740000   
Data/validation/MSP-PODCAST_0396_2225.wav 0 days 0 days 00:00:01.029937500   

                                           F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                                                           
Data/validation/MSP-PODCAST_0744_0066.wav                          32.460304   
Data/validation/MSP-PODCAST_0726_0024.wav                          34.533546   
Data/validation/MSP-PODCAST_0726_0094.wav                          34.142075   
Data/validation/MSP-PODCAST_0653_0154.wav                          39.902229   
Data/validation/MSP-PODCAST_1190_0058.wav                          32.839245   
...                                                                      ...   
Data/validation/MSP-PODCAST_0726_0010.wav                          25.751215   
Data/validation/MSP-PODCAST_1190_0008.wav                          31.448648   
Data/validation/MSP-PODCAST_1185_0019.wav                          31.897440   
Data/validation/MSP-PODCAST_1190_0043.wav                          33.760132   
Data/validation/MSP-PODCAST_0396_2225.wav                          28.339310   

                                           F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                                                                
Data/validation/MSP-PODCAST_0744_0066.wav                                0.122227   
Data/validation/MSP-PODCAST_0726_0024.wav                                0.103023   
Data/validation/MSP-PODCAST_0726_0094.wav                                0.064815   
Data/validation/MSP-PODCAST_0653_0154.wav                                0.098032   
Data/validation/MSP-PODCAST_1190_0058.wav                                0.371941   
...                                                                           ...   
Data/validation/MSP-PODCAST_0726_0010.wav                                0.247039   
Data/validation/MSP-PODCAST_1190_0008.wav                                0.257237   
Data/validation/MSP-PODCAST_1185_0019.wav                                0.175576   
Data/validation/MSP-PODCAST_1190_0043.wav                                0.274083   
Data/validation/MSP-PODCAST_0396_2225.wav                                0.134931   

                                           F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                                                                    
Data/validation/MSP-PODCAST_0744_0066.wav                                   29.499130   
Data/validation/MSP-PODCAST_0726_0024.wav                                   32.856342   
Data/validation/MSP-PODCAST_0726_0094.wav                                   33.418037   
Data/validation/MSP-PODCAST_0653_0154.wav                                   37.958176   
Data/validation/MSP-PODCAST_1190_0058.wav                                   15.883025   
...                                                                               ...   
Data/validation/MSP-PODCAST_0726_0010.wav                                   22.775183   
Data/validation/MSP-PODCAST_1190_0008.wav                                   26.7

In [12]:
file_paths = []
for file in os.listdir(validation_path):
    file_paths.append(os.path.join(validation_path, file))
validation = smile.process_files(file_paths)
validation[validation.isna().any(axis=1)]

/home/changhyun/workspace/ABI_research/env/lib/python3.8/site-packages/opensmile/core/smile.py:308: UserWarning: Segment too short, filling with NaN.
  warnings.warn(


F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                      start  end                                                         
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                      start  end                                                              
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                     NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file                                      start  end                                                                
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                       NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file                                      start  end                                                                   
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                          NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file                                      start  end                                                                     
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                            NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file                                      start  end                                                                    
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                           NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file                                      start  end                                                                      
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                             NaN   

                                                                         ...  \
file                                      start  end                     ...   
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000  ...   

                                                                         slopeUV0-500_sma3nz_amean  \
file                                      start  end                                                 
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:0

In [67]:
indices = []
for index in data.index:
    indices.append(index[16:-4])
# indices
data['podcast'] = indices
data = data.set_index('podcast')
data

/tmp/ipykernel_19296/2402705159.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['podcast'] = indices


start                       end  \
podcast                                                  
MSP-PODCAST_0744_0066 0 days    0 days 00:00:16.630000   
MSP-PODCAST_0726_0024 0 days    0 days 00:00:07.590000   
MSP-PODCAST_0726_0094 0 days    0 days 00:00:05.980000   
MSP-PODCAST_0653_0154 0 days 0 days 00:00:07.560062500   
MSP-PODCAST_1190_0058 0 days    0 days 00:00:05.210000   
...                      ...                       ...   
MSP-PODCAST_0726_0010 0 days    0 days 00:00:15.910000   
MSP-PODCAST_1190_0008 0 days           0 days 00:00:04   
MSP-PODCAST_1185_0019 0 days 0 days 00:00:45.359954649   
MSP-PODCAST_1190_0043 0 days    0 days 00:00:01.740000   
MSP-PODCAST_0396_2225 0 days 0 days 00:00:01.029937500   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0744_0066                          32.460304   
MSP-PODCAST_0726_0024                          34.533546   
MSP-PODCAST_0726_0094                          34.142075   
MSP-PODCAST_0653_0154                          39.902229   
MSP-PODCAST_1190_0058                          32.839245   
...                                                  ...   
MSP-PODCAST_0726_0010                          25.751215   
MSP-PODCAST_1190_0008                          31.448648   
MSP-PODCAST_1185_0019                          31.897440   
MSP-PODCAST_1190_0043                          33.760132   
MSP-PODCAST_0396_2225                          28.339310   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0744_0066                                0.122227   
MSP-PODCAST_0726_0024                                0.103023   
MSP-PODCAST_0726_0094                                0.064815   
MSP-PODCAST_0653_0154                                0.098032   
MSP-PODCAST_1190_0058                                0.371941   
...                                                       ...   
MSP-PODCAST_0726_0010                                0.247039   
MSP-PODCAST_1190_0008                                0.257237   
MSP-PODCAST_1185_0019                                0.175576   
MSP-PODCAST_1190_0043                                0.274083   
MSP-PODCAST_0396_2225                                0.134931   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0744_0066                                   29.499130   
MSP-PODCAST_0726_0024                                   32.856342   
MSP-PODCAST_0726_0094                                   33.418037   
MSP-PODCAST_0653_0154                                   37.958176   
MSP-PODCAST_1190_0058                                   15.883025   
...                                                           ...   
MSP-PODCAST_0726_0010                                   22.775183   
MSP-PODCAST_1190_0008                                   26.765772   
MSP-PODCAST_1185_0019                                   30.385963   
MSP-PODCAST_1190_0043                                   23.972233   
MSP-PODCAST_0396_2225                                   26.839783   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0744_0066                                   31.773491   
MSP-PODCAST_0726_0024                                   33.926315   
MSP-PODCAST_0726_0094                                   34.092876   
MSP-PODCAST_0653_0154                                   40.221134   
MSP-PODCAST_1190_0058                                   36.886265   
...                                                           ...   
MSP-PODCAST_0726_0010                                   24.019869   
MSP-PODCAST_1190_0008                                   34.953243   
MSP-PODCAST_1185_0019                                

In [68]:
li = os.listdir(arouse_path)
list1 = []
for file in li:
    if file[0:21] in validation_files:
        list1.append(file)
list1.sort()
list1
# li[0][0:21] in test_files


['MSP-Conversation_0260_1_001.csv',
 'MSP-Conversation_0260_1_004.csv',
 'MSP-Conversation_0260_1_005.csv',
 'MSP-Conversation_0260_1_006.csv',
 'MSP-Conversation_0260_1_007.csv',
 'MSP-Conversation_0260_1_008.csv',
 'MSP-Conversation_0260_1_009.csv',
 'MSP-Conversation_0260_2_001.csv',
 'MSP-Conversation_0260_2_002.csv',
 'MSP-Conversation_0260_2_005.csv',
 'MSP-Conversation_0260_2_006.csv',
 'MSP-Conversation_0260_2_007.csv',
 'MSP-Conversation_0260_2_010.csv',
 'MSP-Conversation_0260_3_001.csv',
 'MSP-Conversation_0260_3_002.csv',
 'MSP-Conversation_0260_3_003.csv',
 'MSP-Conversation_0260_3_005.csv',
 'MSP-Conversation_0260_3_006.csv',
 'MSP-Conversation_0260_3_007.csv',
 'MSP-Conversation_0260_3_009.csv',
 'MSP-Conversation_0260_3_010.csv',
 'MSP-Conversation_0261_1_001.csv',
 'MSP-Conversation_0261_1_002.csv',
 'MSP-Conversation_0261_1_003.csv',
 'MSP-Conversation_0261_1_005.csv',
 'MSP-Conversation_0261_1_006.csv',
 'MSP-Conversation_0261_1_007.csv',
 'MSP-Conversation_0261_1_00

In [69]:
arouse_annotation = {}
for elem in list1:
    if elem[0:23] in arouse_annotation.keys():
        arouse_annotation.get(elem[0:23]).append(elem)
    else:
        arouse_annotation[elem[0:23]] = [elem]
arouse_annotation

{'MSP-Conversation_0260_1': ['MSP-Conversation_0260_1_001.csv',
  'MSP-Conversation_0260_1_004.csv',
  'MSP-Conversation_0260_1_005.csv',
  'MSP-Conversation_0260_1_006.csv',
  'MSP-Conversation_0260_1_007.csv',
  'MSP-Conversation_0260_1_008.csv',
  'MSP-Conversation_0260_1_009.csv'],
 'MSP-Conversation_0260_2': ['MSP-Conversation_0260_2_001.csv',
  'MSP-Conversation_0260_2_002.csv',
  'MSP-Conversation_0260_2_005.csv',
  'MSP-Conversation_0260_2_006.csv',
  'MSP-Conversation_0260_2_007.csv',
  'MSP-Conversation_0260_2_010.csv'],
 'MSP-Conversation_0260_3': ['MSP-Conversation_0260_3_001.csv',
  'MSP-Conversation_0260_3_002.csv',
  'MSP-Conversation_0260_3_003.csv',
  'MSP-Conversation_0260_3_005.csv',
  'MSP-Conversation_0260_3_006.csv',
  'MSP-Conversation_0260_3_007.csv',
  'MSP-Conversation_0260_3_009.csv',
  'MSP-Conversation_0260_3_010.csv'],
 'MSP-Conversation_0261_1': ['MSP-Conversation_0261_1_001.csv',
  'MSP-Conversation_0261_1_002.csv',
  'MSP-Conversation_0261_1_003.csv',
 

In [70]:
li2 = os.listdir(valence_path)
list2 = []
for file in li2:
    if file[0:21] in validation_files:
        list2.append(file)
list2.sort()
list2

['MSP-Conversation_0260_1_001.csv',
 'MSP-Conversation_0260_1_002.csv',
 'MSP-Conversation_0260_1_003.csv',
 'MSP-Conversation_0260_1_005.csv',
 'MSP-Conversation_0260_1_006.csv',
 'MSP-Conversation_0260_1_007.csv',
 'MSP-Conversation_0260_1_011.csv',
 'MSP-Conversation_0260_2_001.csv',
 'MSP-Conversation_0260_2_002.csv',
 'MSP-Conversation_0260_2_003.csv',
 'MSP-Conversation_0260_2_005.csv',
 'MSP-Conversation_0260_2_006.csv',
 'MSP-Conversation_0260_2_007.csv',
 'MSP-Conversation_0260_3_001.csv',
 'MSP-Conversation_0260_3_002.csv',
 'MSP-Conversation_0260_3_003.csv',
 'MSP-Conversation_0260_3_005.csv',
 'MSP-Conversation_0260_3_006.csv',
 'MSP-Conversation_0260_3_008.csv',
 'MSP-Conversation_0261_1_001.csv',
 'MSP-Conversation_0261_1_005.csv',
 'MSP-Conversation_0261_1_006.csv',
 'MSP-Conversation_0261_1_007.csv',
 'MSP-Conversation_0261_1_009.csv',
 'MSP-Conversation_0261_1_011.csv',
 'MSP-Conversation_0261_2_001.csv',
 'MSP-Conversation_0261_2_002.csv',
 'MSP-Conversation_0261_2_00

In [71]:
li3 = os.listdir(dominance_path)
list3 = []
for file in li3:
    if file[0:21] in validation_files:
        list3.append(file)
list3.sort()
list3

['MSP-Conversation_0260_1_001.csv',
 'MSP-Conversation_0260_1_002.csv',
 'MSP-Conversation_0260_1_005.csv',
 'MSP-Conversation_0260_1_006.csv',
 'MSP-Conversation_0260_1_007.csv',
 'MSP-Conversation_0260_1_009.csv',
 'MSP-Conversation_0260_1_011.csv',
 'MSP-Conversation_0260_2_001.csv',
 'MSP-Conversation_0260_2_002.csv',
 'MSP-Conversation_0260_2_005.csv',
 'MSP-Conversation_0260_2_006.csv',
 'MSP-Conversation_0260_2_007.csv',
 'MSP-Conversation_0260_2_010.csv',
 'MSP-Conversation_0260_3_001.csv',
 'MSP-Conversation_0260_3_005.csv',
 'MSP-Conversation_0260_3_006.csv',
 'MSP-Conversation_0260_3_007.csv',
 'MSP-Conversation_0260_3_008.csv',
 'MSP-Conversation_0260_3_009.csv',
 'MSP-Conversation_0261_1_001.csv',
 'MSP-Conversation_0261_1_005.csv',
 'MSP-Conversation_0261_1_006.csv',
 'MSP-Conversation_0261_1_007.csv',
 'MSP-Conversation_0261_1_008.csv',
 'MSP-Conversation_0261_1_010.csv',
 'MSP-Conversation_0261_1_011.csv',
 'MSP-Conversation_0261_2_001.csv',
 'MSP-Conversation_0261_2_00

In [72]:
valence_annotation = {}
dominance_annotation = {}
for elem in list2:
    if elem[0:23] in valence_annotation.keys():
        valence_annotation.get(elem[0:23]).append(elem)
    else:
        valence_annotation[elem[0:23]] = [elem]
for elem in list3:
    if elem[0:23] in dominance_annotation.keys():
        dominance_annotation.get(elem[0:23]).append(elem)
    else:
        dominance_annotation[elem[0:23]] = [elem]
valence_annotation

{'MSP-Conversation_0260_1': ['MSP-Conversation_0260_1_001.csv',
  'MSP-Conversation_0260_1_002.csv',
  'MSP-Conversation_0260_1_003.csv',
  'MSP-Conversation_0260_1_005.csv',
  'MSP-Conversation_0260_1_006.csv',
  'MSP-Conversation_0260_1_007.csv',
  'MSP-Conversation_0260_1_011.csv'],
 'MSP-Conversation_0260_2': ['MSP-Conversation_0260_2_001.csv',
  'MSP-Conversation_0260_2_002.csv',
  'MSP-Conversation_0260_2_003.csv',
  'MSP-Conversation_0260_2_005.csv',
  'MSP-Conversation_0260_2_006.csv',
  'MSP-Conversation_0260_2_007.csv'],
 'MSP-Conversation_0260_3': ['MSP-Conversation_0260_3_001.csv',
  'MSP-Conversation_0260_3_002.csv',
  'MSP-Conversation_0260_3_003.csv',
  'MSP-Conversation_0260_3_005.csv',
  'MSP-Conversation_0260_3_006.csv',
  'MSP-Conversation_0260_3_008.csv'],
 'MSP-Conversation_0261_1': ['MSP-Conversation_0261_1_001.csv',
  'MSP-Conversation_0261_1_005.csv',
  'MSP-Conversation_0261_1_006.csv',
  'MSP-Conversation_0261_1_007.csv',
  'MSP-Conversation_0261_1_009.csv',
 

In [73]:
dominance_annotation

{'MSP-Conversation_0260_1': ['MSP-Conversation_0260_1_001.csv',
  'MSP-Conversation_0260_1_002.csv',
  'MSP-Conversation_0260_1_005.csv',
  'MSP-Conversation_0260_1_006.csv',
  'MSP-Conversation_0260_1_007.csv',
  'MSP-Conversation_0260_1_009.csv',
  'MSP-Conversation_0260_1_011.csv'],
 'MSP-Conversation_0260_2': ['MSP-Conversation_0260_2_001.csv',
  'MSP-Conversation_0260_2_002.csv',
  'MSP-Conversation_0260_2_005.csv',
  'MSP-Conversation_0260_2_006.csv',
  'MSP-Conversation_0260_2_007.csv',
  'MSP-Conversation_0260_2_010.csv'],
 'MSP-Conversation_0260_3': ['MSP-Conversation_0260_3_001.csv',
  'MSP-Conversation_0260_3_005.csv',
  'MSP-Conversation_0260_3_006.csv',
  'MSP-Conversation_0260_3_007.csv',
  'MSP-Conversation_0260_3_008.csv',
  'MSP-Conversation_0260_3_009.csv'],
 'MSP-Conversation_0261_1': ['MSP-Conversation_0261_1_001.csv',
  'MSP-Conversation_0261_1_005.csv',
  'MSP-Conversation_0261_1_006.csv',
  'MSP-Conversation_0261_1_007.csv',
  'MSP-Conversation_0261_1_008.csv',
 

In [74]:
segment_path = os.path.relpath('./MSP Data/Time Labels/segments.json')
f = open(segment_path, 'r')
timing_data = json.load(f)
temp = data.copy()

In [75]:
def mean(list):
    return sum(list) / len(list)

In [76]:
for key in timing_data:
    conv_part = timing_data[key]['Conversation_Part']
    if key not in temp.index:
        continue
    start = timing_data[key]['Start_Time']
    end = timing_data[key]['End_Time']
    arousal = []
    valence = []
    dominance = []
    a_annotations = arouse_annotation.get(conv_part)
    v_annotations = valence_annotation.get(conv_part)
    d_annotations = dominance_annotation.get(conv_part)
    for anno in a_annotations:
        pp = os.path.join(arouse_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'arousal'])
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        arousal.append(df['arousal'].mean())
    
    for anno in v_annotations:
        pp = os.path.join(valence_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'valence'])
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        valence.append(df['valence'].mean())
    for anno in d_annotations:
        pp = os.path.join(dominance_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'dominance'])
        if(anno == 'MSP-Conversation_0047_2_001.csv'):
            df = df.reset_index()
            df = df.drop(columns=['dominance'])
            df.columns = ['time', 'dominance']
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        
        dominance.append(df['dominance'].mean())
    
    score_arousal = mean(arousal)
    score_valence = mean(valence)
    score_dominance = mean(dominance)
    
    temp.loc[key, 'arousal'] = score_arousal
    temp.loc[key, 'valence'] = score_valence
    temp.loc[key, 'dominance'] = score_dominance
    
temp

start                       end  \
podcast                                                  
MSP-PODCAST_0744_0066 0 days    0 days 00:00:16.630000   
MSP-PODCAST_0726_0024 0 days    0 days 00:00:07.590000   
MSP-PODCAST_0726_0094 0 days    0 days 00:00:05.980000   
MSP-PODCAST_0653_0154 0 days 0 days 00:00:07.560062500   
MSP-PODCAST_1190_0058 0 days    0 days 00:00:05.210000   
...                      ...                       ...   
MSP-PODCAST_0726_0010 0 days    0 days 00:00:15.910000   
MSP-PODCAST_1190_0008 0 days           0 days 00:00:04   
MSP-PODCAST_1185_0019 0 days 0 days 00:00:45.359954649   
MSP-PODCAST_1190_0043 0 days    0 days 00:00:01.740000   
MSP-PODCAST_0396_2225 0 days 0 days 00:00:01.029937500   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0744_0066                          32.460304   
MSP-PODCAST_0726_0024                          34.533546   
MSP-PODCAST_0726_0094                          34.142075   
MSP-PODCAST_0653_0154                          39.902229   
MSP-PODCAST_1190_0058                          32.839245   
...                                                  ...   
MSP-PODCAST_0726_0010                          25.751215   
MSP-PODCAST_1190_0008                          31.448648   
MSP-PODCAST_1185_0019                          31.897440   
MSP-PODCAST_1190_0043                          33.760132   
MSP-PODCAST_0396_2225                          28.339310   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0744_0066                                0.122227   
MSP-PODCAST_0726_0024                                0.103023   
MSP-PODCAST_0726_0094                                0.064815   
MSP-PODCAST_0653_0154                                0.098032   
MSP-PODCAST_1190_0058                                0.371941   
...                                                       ...   
MSP-PODCAST_0726_0010                                0.247039   
MSP-PODCAST_1190_0008                                0.257237   
MSP-PODCAST_1185_0019                                0.175576   
MSP-PODCAST_1190_0043                                0.274083   
MSP-PODCAST_0396_2225                                0.134931   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0744_0066                                   29.499130   
MSP-PODCAST_0726_0024                                   32.856342   
MSP-PODCAST_0726_0094                                   33.418037   
MSP-PODCAST_0653_0154                                   37.958176   
MSP-PODCAST_1190_0058                                   15.883025   
...                                                           ...   
MSP-PODCAST_0726_0010                                   22.775183   
MSP-PODCAST_1190_0008                                   26.765772   
MSP-PODCAST_1185_0019                                   30.385963   
MSP-PODCAST_1190_0043                                   23.972233   
MSP-PODCAST_0396_2225                                   26.839783   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0744_0066                                   31.773491   
MSP-PODCAST_0726_0024                                   33.926315   
MSP-PODCAST_0726_0094                                   34.092876   
MSP-PODCAST_0653_0154                                   40.221134   
MSP-PODCAST_1190_0058                                   36.886265   
...                                                           ...   
MSP-PODCAST_0726_0010                                   24.019869   
MSP-PODCAST_1190_0008                                   34.953243   
MSP-PODCAST_1185_0019                                

In [77]:
temp[temp.isnull().any(axis=1)]

Empty DataFrame
Columns: [start, end, F0semitoneFrom27.5Hz_sma3nz_amean, F0semitoneFrom27.5Hz_sma3nz_stddevNorm, F0semitoneFrom27.5Hz_sma3nz_percentile20.0, F0semitoneFrom27.5Hz_sma3nz_percentile50.0, F0semitoneFrom27.5Hz_sma3nz_percentile80.0, F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2, F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope, F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope, loudness_sma3_amean, loudness_sma3_stddevNorm, loudness_sma3_percentile20.0, loudness_sma3_percentile50.0, loudness_sma3_percentile80.0, loudness_sma3_pctlrange0-2, loudness_sma3_meanRisingSlope, loudness_sma3_stddevRisingSlope, loudness_sma3_meanFallingSlope, loudness_sma3_stddevFallingSlope, spectralFlux_sma3_amean, spectralFlux_sma3_stddevNorm, mfcc1_sma3_amean, mfcc1_sma3_stddevNorm, mfcc2_sma3_amean, mfcc2_sma3_stddevNorm, mfcc3_sma3_amean, mfcc3_sma3_stddevNorm, mfcc4_sma3_amean, mfcc4_sma3_stddevNorm, jitterLocal_sma3nz_amean, jitterLocal_sma3nz_stddevNorm, shimmerLocaldB_sma3nz_amean, shimmerLocaldB_sma3nz_stddevNorm, HNRdBACF_sma3nz_amean, HNRdBACF_sma3nz_stddevNorm, logRelF0-H1-H2_sma3nz_amean, logRelF0-H1-H2_sma3nz_stddevNorm, logRelF0-H1-A3_sma3nz_amean, logRelF0-H1-A3_sma3nz_stddevNorm, F1frequency_sma3nz_amean, F1frequency_sma3nz_stddevNorm, F1bandwidth_sma3nz_amean, F1bandwidth_sma3nz_stddevNorm, F1amplitudeLogRelF0_sma3nz_amean, F1amplitudeLogRelF0_sma3nz_stddevNorm, F2frequency_sma3nz_amean, F2frequency_sma3nz_stddevNorm, F2bandwidth_sma3nz_amean, F2bandwidth_sma3nz_stddevNorm, F2amplitudeLogRelF0_sma3nz_amean, F2amplitudeLogRelF0_sma3nz_stddevNorm, F3frequency_sma3nz_amean, F3frequency_sma3nz_stddevNorm, F3bandwidth_sma3nz_amean, F3bandwidth_sma3nz_stddevNorm, F3amplitudeLogRelF0_sma3nz_amean, F3amplitudeLogRelF0_sma3nz_stddevNorm, alphaRatioV_sma3nz_amean, alphaRatioV_sma3nz_stddevNorm, hammarbergIndexV_sma3nz_amean, hammarbergIndexV_sma3nz_stddevNorm, slopeV0-500_sma3nz_amean, slopeV0-500_sma3nz_stddevNorm, slopeV500-1500_sma3nz_amean, slopeV500-1500_sma3nz_stddevNorm, spectralFluxV_sma3nz_amean, spectralFluxV_sma3nz_stddevNorm, mfcc1V_sma3nz_amean, mfcc1V_sma3nz_stddevNorm, mfcc2V_sma3nz_amean, mfcc2V_sma3nz_stddevNorm, mfcc3V_sma3nz_amean, mfcc3V_sma3nz_stddevNorm, mfcc4V_sma3nz_amean, mfcc4V_sma3nz_stddevNorm, alphaRatioUV_sma3nz_amean, hammarbergIndexUV_sma3nz_amean, slopeUV0-500_sma3nz_amean, slopeUV500-1500_sma3nz_amean, spectralFluxUV_sma3nz_amean, loudnessPeaksPerSec, VoicedSegmentsPerSec, MeanVoicedSegmentLengthSec, StddevVoicedSegmentLengthSec, MeanUnvoicedSegmentLength, StddevUnvoicedSegmentLength, equivalentSoundLevel_dBp, arousal, valence, dominance]
Index: []

[0 rows x 93 columns]

In [78]:
temp = temp.sort_values(by='podcast')
temp

start                       end  \
podcast                                                  
MSP-PODCAST_0260_0386 0 days    0 days 00:00:03.550000   
MSP-PODCAST_0260_0389 0 days    0 days 00:00:03.310000   
MSP-PODCAST_0260_0390 0 days    0 days 00:00:03.290000   
MSP-PODCAST_0260_0391 0 days 0 days 00:00:05.510062500   
MSP-PODCAST_0260_0394 0 days    0 days 00:00:03.310000   
...                      ...                       ...   
MSP-PODCAST_1191_0022 0 days    0 days 00:00:43.170000   
MSP-PODCAST_1191_0023 0 days    0 days 00:00:58.870000   
MSP-PODCAST_1191_0024 0 days    0 days 00:00:07.520000   
MSP-PODCAST_1191_0025 0 days 0 days 00:00:30.630062500   
MSP-PODCAST_1191_0026 0 days 0 days 00:01:32.359937500   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0260_0386                          34.466839   
MSP-PODCAST_0260_0389                          29.448351   
MSP-PODCAST_0260_0390                          36.067398   
MSP-PODCAST_0260_0391                          28.304644   
MSP-PODCAST_0260_0394                          27.772774   
...                                                  ...   
MSP-PODCAST_1191_0022                          31.804155   
MSP-PODCAST_1191_0023                          33.649513   
MSP-PODCAST_1191_0024                          32.677185   
MSP-PODCAST_1191_0025                          36.688194   
MSP-PODCAST_1191_0026                          34.225868   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0260_0386                                0.074569   
MSP-PODCAST_0260_0389                                0.078086   
MSP-PODCAST_0260_0390                                0.111166   
MSP-PODCAST_0260_0391                                0.084098   
MSP-PODCAST_0260_0394                                0.106502   
...                                                       ...   
MSP-PODCAST_1191_0022                                0.168613   
MSP-PODCAST_1191_0023                                0.108885   
MSP-PODCAST_1191_0024                                0.110571   
MSP-PODCAST_1191_0025                                0.102438   
MSP-PODCAST_1191_0026                                0.195363   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0260_0386                                   32.814217   
MSP-PODCAST_0260_0389                                   27.304937   
MSP-PODCAST_0260_0390                                   32.458305   
MSP-PODCAST_0260_0391                                   26.575451   
MSP-PODCAST_0260_0394                                   24.562349   
...                                                           ...   
MSP-PODCAST_1191_0022                                   27.741779   
MSP-PODCAST_1191_0023                                   30.838093   
MSP-PODCAST_1191_0024                                   30.151619   
MSP-PODCAST_1191_0025                                   34.354458   
MSP-PODCAST_1191_0026                                   29.493553   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0260_0386                                   34.846764   
MSP-PODCAST_0260_0389                                   29.390343   
MSP-PODCAST_0260_0390                                   35.660553   
MSP-PODCAST_0260_0391                                   28.722303   
MSP-PODCAST_0260_0394                                   28.643770   
...                                                           ...   
MSP-PODCAST_1191_0022                                   30.843397   
MSP-PODCAST_1191_0023                                   33.249950   
MSP-PODCAST_1191_0024                                

In [79]:

temp.to_csv('validation.csv')

In [115]:
# These values are the total number of each sample
print('Total: ', len(timing_data))
print('Train: ', len(os.listdir(train_path)))
print('Validation: ', len(os.listdir(validation_path)))
print('Test: ', len(os.listdir(test_path)))

Total:  5036
Train:  2783
Validation:  750
Test:  1503


However, there are missing values that openSmile cannot fetch acoustic features from certain audio segments:

Validation dataset: MSP-PODCAST_1188_0020 (Segment name) (Total: 1)
Train dataset: 81 segments in MSP-Conversation_0.153_2.wav were missing (Total: 81) => In segments.json, start time and end time are exactly the same which is 0-length audio segments.

Extraction results:
- Train: 2702
- Validation: 749
- Test: 1503


In [134]:
temp.shape

(2702, 93)

In [78]:
path = os.path.relpath('failed.txt')

with open(path) as f:
    lines = f.read().splitlines()
lines = sorted(lines)
lines

['Data/test/MSP-PODCAST_0047_0279.wav',
 'Data/test/MSP-PODCAST_0114_0281.wav',
 'Data/test/MSP-PODCAST_0114_0291.wav',
 'Data/test/MSP-PODCAST_0167_0128.wav',
 'Data/test/MSP-PODCAST_0167_0152.wav',
 'Data/test/MSP-PODCAST_0167_0205.wav',
 'Data/test/MSP-PODCAST_0190_0025.wav',
 'Data/test/MSP-PODCAST_0190_0081.wav',
 'Data/test/MSP-PODCAST_0202_0018.wav',
 'Data/test/MSP-PODCAST_0218_0190.wav',
 'Data/test/MSP-PODCAST_0281_0022.wav',
 'Data/test/MSP-PODCAST_0281_0028.wav',
 'Data/test/MSP-PODCAST_0281_0033.wav',
 'Data/test/MSP-PODCAST_0281_0034.wav',
 'Data/test/MSP-PODCAST_0295_0463.wav',
 'Data/test/MSP-PODCAST_0381_0276.wav',
 'Data/test/MSP-PODCAST_0381_0278.wav',
 'Data/test/MSP-PODCAST_0381_0280.wav',
 'Data/test/MSP-PODCAST_0381_0283.wav',
 'Data/test/MSP-PODCAST_0381_0285.wav',
 'Data/test/MSP-PODCAST_0381_0287.wav',
 'Data/test/MSP-PODCAST_0381_0288.wav',
 'Data/test/MSP-PODCAST_0381_0290.wav',
 'Data/test/MSP-PODCAST_0381_0292.wav',
 'Data/test/MSP-PODCAST_0381_0294.wav',


In [81]:
f = open("failed_sorted.txt", "a")
for line in lines:
    f.write(line+'\n')
f.close()
